File: cdk/src/main/java/com/myorg/DockerFargateStack.java
package com.myorg;
import java.util.Collections;
import java.util.Map;

import software.amazon.awscdk.Stack;
import software.amazon.awscdk.StackProps;
import software.amazon.awscdk.Tags;
import software.amazon.awscdk.services.ec2.SecurityGroup;
import software.amazon.awscdk.services.ec2.Vpc;
import software.amazon.awscdk.services.ecs.Cluster;
import software.amazon.awscdk.services.ecs.ContainerDefinitionOptions;
import software.amazon.awscdk.services.ecs.ContainerImage;
import software.amazon.awscdk.services.ecs.FargateTaskDefinition;
import software.amazon.awscdk.services.ecs.PortMapping;
import software.amazon.awscdk.services.ecs.Secret;
import software.amazon.awscdk.services.ecs.patterns.ApplicationLoadBalancedFargateService;
import software.amazon.awscdk.services.elasticloadbalancingv2.AddApplicationTargetsProps;
import software.amazon.awscdk.services.elasticloadbalancingv2.ApplicationListener;
import software.amazon.awscdk.services.elasticloadbalancingv2.ApplicationLoadBalancer;
import software.amazon.awscdk.services.elasticloadbalancingv2.BaseApplicationListenerProps;
import software.amazon.awscdk.services.secretsmanager.ISecret;
import software.constructs.Construct;
import org.apache.commons.lang.StringUtils;
import software.amazon.awscdk.services.ecs.patterns.ApplicationLoadBalancedTaskImageOptions;
import java.util.Map;
import java.util.HashMap;

public class DockerFargateStack extends Stack {
	private static final String STACK_NAME_PREFIX = "STACK_NAME_PREFIX";
	private static final String ID_SUFFIX = "-DockerFargateStack";
	private static final String VPC_SUFFIX = "-FargateVPC";
	private static final String CLUSTER_SUFFIX = "-Cluster";
	private static final String SERVICE_SUFFIX = "-Service";
	private static final String DOCKER_IMAGE_NAME = "DOCKER_IMAGE";
	private static final String COST_CENTER = "COST_CENTER";
	private static final String COST_CENTER_TAG_NAME = "CostCenter";
	private static final String PORT_NUMBER = "PORT";
	private static final String SECRETS_MANAGER_SYNAPSE_AUTH_TOKEN_KEY = "synapse-auth-token";
	private static final String SECRETS_MANAGER_GOOGLE_AUTH_JSON_KEY = "google-auth-json";
	private static final String SECRETS_MANAGER_ENV_PREFIX = "secrets-manager-";
	
	// the following are the secret names as they will appear inside the container
	private static final String SECRET_ENV_NAME_SYNAPSE_AUTH_TOKEN =
			SECRETS_MANAGER_ENV_PREFIX+SECRETS_MANAGER_SYNAPSE_AUTH_TOKEN_KEY;
	private static final String SECRET_ENV_NAME_GOOGLE_AUTH_JSON =
			SECRETS_MANAGER_ENV_PREFIX+SECRETS_MANAGER_GOOGLE_AUTH_JSON_KEY;
	
	private static final String CONTAINER_ENV = "CONTAINER_ENV"; // name of env passed from GitHub action
	private static final String ENV_NAME = "ENV"; // name of env as seen in container
	
	private String id;
	
	private static String getRequiredEnv(String envName) {
		String result = System.getenv(envName);
		if (StringUtils.isEmpty(result)) {
			throw new IllegalStateException(envName+" is required.");
		}
		return result;
	}
	
	private static String getId() {
		return getRequiredEnv(STACK_NAME_PREFIX)+ID_SUFFIX;
	}
	
	private static String getVpcName() {
		return getRequiredEnv(STACK_NAME_PREFIX)+VPC_SUFFIX;
	}
	
	private static String getClusterName() {
		return getRequiredEnv(STACK_NAME_PREFIX)+CLUSTER_SUFFIX;
	}
	
	private static String getServiceName() {
		return getRequiredEnv(STACK_NAME_PREFIX)+SERVICE_SUFFIX;
	}
	
	private static String getDockerImageName() {
		return getRequiredEnv(DOCKER_IMAGE_NAME);
	}
	
	private static String getCostCenter() {
		return getRequiredEnv(COST_CENTER);
	}
	
	private static int getPort() {
		return Integer.parseInt(getRequiredEnv(PORT_NUMBER));
	}
	
	private static String getSecretsManagerSynapseAuthTokenKey() {
		return getRequiredEnv(STACK_NAME_PREFIX)+"/"+
				System.getenv(SECRETS_MANAGER_SYNAPSE_AUTH_TOKEN_KEY);
	}
	
	private static String getSecretsManagerGoogleAuthJSONKey() {
		return getRequiredEnv(STACK_NAME_PREFIX)+"/"+
				System.getenv(SECRETS_MANAGER_GOOGLE_AUTH_JSON_KEY);
	}
	
	private static String getContainerEnv() {
		return System.getenv(CONTAINER_ENV);
	}
	
	public DockerFargateStack(final Construct scope) {
		this(scope, null);
	}
	
	private Secret createSecret(String name) {
		ISecret isecret = software.amazon.awscdk.services.secretsmanager.
				Secret.fromSecretNameV2(this, id, name);
		return Secret.fromSecretsManager(isecret);
	}


	public DockerFargateStack(final Construct scope, final StackProps props) {
		super(scope, getId(), props);
		this.id=getId();

		Vpc vpc = Vpc.Builder.create(this, getVpcName())
				.maxAzs(2)
				.build();

		Cluster cluster = Cluster.Builder.create(this, getClusterName())
				.vpc(vpc).build();
		
		ApplicationLoadBalancedTaskImageOptions.Builder imageOptionsBuilder = 
				ApplicationLoadBalancedTaskImageOptions.builder()
				.image(ContainerImage.fromRegistry(getDockerImageName()))
				.containerPort(getPort());
		
		Map<String, Secret> secrets = 
				new HashMap<>();
		secrets.put(SECRET_ENV_NAME_SYNAPSE_AUTH_TOKEN, 
				createSecret(getSecretsManagerSynapseAuthTokenKey()));
		secrets.put(SECRET_ENV_NAME_GOOGLE_AUTH_JSON, 
				createSecret(getSecretsManagerGoogleAuthJSONKey()));
		imageOptionsBuilder.secrets(secrets);
		
		// optionally pass in a plain-text (not secret) environment variable
		String containerEnvVar = getContainerEnv();
		if (StringUtils.isNotEmpty(containerEnvVar)) {
			Map<String,String> envVars = Collections.singletonMap(ENV_NAME, containerEnvVar);
			imageOptionsBuilder.environment(envVars);
		}
		ApplicationLoadBalancedTaskImageOptions imageOptions = 
				imageOptionsBuilder.build();

		// Create a load-balanced Fargate service and make it public
		ApplicationLoadBalancedFargateService albfs = 
				ApplicationLoadBalancedFargateService.Builder.create(this, getServiceName())
				.cluster(cluster)           // Required
				.cpu(256)                   // Default is 256
				.desiredCount(2)            // Default is 1
				.taskImageOptions(imageOptions)
				.memoryLimitMiB(1024)       // Default is 512
				.publicLoadBalancer(true)
				.build();   // Default is false
		
		Tags.of(albfs).add(COST_CENTER_TAG_NAME, getCostCenter());
	}
}

File: cdk/docker_fargate/docker_fargate_stack.py

from aws_cdk import (Stack,
	aws_ec2 as ec2, aws_ecs as ecs,
	aws_ecs_patterns as ecs_patterns,
	aws_ssm as ssm,
	Tags)

from constructs import Construct
import os
import aws_cdk.aws_secretsmanager as sm

STACK_NAME_PREFIX = "STACK_NAME_PREFIX"
ID_SUFFIX = "-DockerFargateStack"
VPC_SUFFIX = "-FargateVPC"
CLUSTER_SUFFIX = "-Cluster"
SERVICE_SUFFIX = "-Service"
DOCKER_IMAGE_NAME = "DOCKER_IMAGE"
COST_CENTER = "COST_CENTER"
COST_CENTER_TAG_NAME = "CostCenter"
PORT_NUMBER = "PORT"
SECRETS_MANAGER_SYNAPSE_AUTH_TOKEN_KEY = "SECRETS_MANAGER_SYNAPSE_AUTH_TOKEN_KEY"
SECRETS_MANAGER_GOOGLE_AUTH_JSON_KEY = "SECRETS_MANAGER_GOOGLE_AUTH_JSON_KEY"
SECRETS_MANAGER_ENV_PREFIX = "secrets-manager-"

def get_required_env(name: str) -> str:
        value = os.getenv(name)
        if value is None or len(value)==0:
            raise Exception(f'{name} is required.')
        return value
		
def create_id() -> str:
        return get_required_env(STACK_NAME_PREFIX)+ID_SUFFIX

# the following are the secret names as they will appear inside the container
SECRET_ENV_NAME_SYNAPSE_AUTH_TOKEN = \
  SECRETS_MANAGER_ENV_PREFIX+SECRETS_MANAGER_SYNAPSE_AUTH_TOKEN_KEY
SECRET_ENV_NAME_GOOGLE_AUTH_JSON = \
  SECRETS_MANAGER_ENV_PREFIX+SECRETS_MANAGER_GOOGLE_AUTH_JSON_KEY

CONTAINER_ENV = "CONTAINER_ENV" # name of env passed from GitHub action
ENV_NAME = "ENV" # name of env as seen in container

def get_vpc_name() -> str:
    return get_required_env(STACK_NAME_PREFIX)+VPC_SUFFIX

def get_cluster_name() -> str:
    return get_required_env(STACK_NAME_PREFIX)+CLUSTER_SUFFIX

def get_service_name() -> str:
    return get_required_env(STACK_NAME_PREFIX)+SERVICE_SUFFIX

def get_docker_image_name():
    return get_required_env(DOCKER_IMAGE_NAME)

def get_cost_center() -> str:
    return get_required_env(COST_CENTER)

def get_port() -> int:
    return int(get_required_env(PORT_NUMBER))

def get_secrets_manager_synapse_auth_token_key() -> str:
    return get_required_env(SECRETS_MANAGER_SYNAPSE_AUTH_TOKEN_KEY)

def get_secrets_manager_google_auth_json_key() -> str:
    return get_required_env(SECRETS_MANAGER_GOOGLE_AUTH_JSON_KEY)

def get_container_env() -> str:
    return os.getenv(CONTAINER_ENV)
          
def create_secret(scope: Construct, name: str) -> str:
    isecret = sm.Secret.from_secret_name_v2(scope, name, name)
    return ecs.Secret.from_secrets_manager(isecret)
    # see also: https://docs.aws.amazon.com/cdk/api/v1/python/aws_cdk.aws_ecs/Secret.html
    # see also: ecs.Secret.from_ssm_parameter(ssm.IParameter(parameter_name=name))

class DockerFargateStack(Stack):

    def __init__(self, scope: Construct, **kwargs) -> None:
    	stack_id = create_id()
    	super().__init__(scope, stack_id, **kwargs)

    	vpc = ec2.Vpc(self, get_vpc_name(), max_azs=2)

    	cluster = ecs.Cluster(self, get_cluster_name(), vpc=vpc)
        
    	secrets = {
        	SECRET_ENV_NAME_SYNAPSE_AUTH_TOKEN: create_secret(self, get_secrets_manager_synapse_auth_token_key()),
        	SECRET_ENV_NAME_GOOGLE_AUTH_JSON:  create_secret(self, get_secrets_manager_google_auth_json_key())
        }
        
    	env_vars = {}
    	container_env = get_container_env()
    	if container_env is not None:
    	    env_vars[ENV_NAME]=container_env
        
    	task_image_options = ecs_patterns.ApplicationLoadBalancedTaskImageOptions(
    	    	image=ecs.ContainerImage.from_registry(get_docker_image_name()),
    	    	environment=env_vars,
    	    	secrets = secrets,
    	    	container_port = get_port())

		#
		# for options to pass to ApplicationLoadBalancedTaskImageOptions see:
		# https://docs.aws.amazon.com/cdk/api/v1/python/aws_cdk.aws_ecs_patterns/ApplicationLoadBalancedTaskImageOptions.html#aws_cdk.aws_ecs_patterns.ApplicationLoadBalancedTaskImageOptions
		#
    	albfs = ecs_patterns.ApplicationLoadBalancedFargateService(self, get_service_name(),
            cluster=cluster,            # Required
            cpu=256,                    # Default is 256
            desired_count=2,            # Default is 1
            task_image_options=task_image_options,
            memory_limit_mib=1024,      # Default is 512
            public_load_balancer=True)  # Default is False

    	Tags.of(albfs).add(COST_CENTER_TAG_NAME, get_cost_center())
